src/main/java/org/passay/HaveIBeenPwnedRule.java

```diff
@@ -0,0 +1,227 @@
+  {
+    this(appName, DEFAULT_URL);
+  }
+
+
+  /**
+   * Create the rule, appName is required by the
+   * <a href="https://haveibeenpwned.com/API/v3#UserAgent">API</a>.
+   *
+   * @param appName must not be null
+   * @param address the URL must end with a <code>/</code>.
+   */
+  public HaveIBeenPwnedRule(final String appName, final String address)
+  {
+    if (appName == null) {
+      throw new IllegalArgumentException("appName cannot be null");
+    }
+    applicationName = appName;
+    if (!address.endsWith("/")) {
+      throw new IllegalArgumentException("address must end with '/'");
+    }
+    try {
+      apiUrl = new URL(address);
+    } catch (MalformedURLException e) {
+      throw new IllegalArgumentException(e);
+    }
+  }
+
+
+  /**
+   * Whether passwords found in the API should be considered valid.
+   *
+   * @param allow false: the rule does not allow previously pwned passwords,
+   *              true: pwned passwords are allowed, but the number of matches is returned in the result.
+   */
+  public void setAllowExposed(final boolean allow)
+  {
+    allowExposed = allow;
+  }
+
+
+  /**
+   * maximum Duration for connecting to the API.
+   *
+   * @param timeout for connecting.
+   */
+  public void setConnectTimeout(final Duration timeout)
+  {
+    connectTimeout = timeout;
+  }
+
+
+  /**
+   * maximum Duration for reading from the API.
+   *
+   * @param timeout for reading.
+   */
+  public void setReadTimeout(final Duration timeout)
+  {
+    readTimeout = timeout;
+  }
+
+
+  /**
+   * If an exception occurs during accessing the api, the password will be allowed, if set to true.
+   *
+   * @param allow true: if the API is not accessible, any password is accepted.
+   *              false: Default, API must answer in time to allow the password.
+   */
+  public void setAllowOnException(final boolean allow)
+  {
+    allowOnException = allow;
+  }
+
+
+  @Override
+  public RuleResult validate(final PasswordData passwordData)
+  {
+    final String hexDigest = getHexDigest(passwordData);
+    try (LineNumberReader lnr = openApiConnectionForRange(hexDigest.substring(0, PREFIX_LENGTH))) {
+      return searchResponse(hexDigest, lnr);
+    } catch (IOException e) {
+      return new RuleResult(allowOnException,
+        new RuleResultDetail(IO_ERROR_CODE, Collections.singletonMap("url", apiUrl)));
+    }
+  }
+
+
+  /**
+   * Reads the supplied reader line by line until a match is found against the supplied hex digest.
+   *
+   * @param hexDigest to match
+   * @param reader to read
+   *
+   * @return rule result whose validity is determined by whether a match was found
+   *
+   * @throws IOException if an error occurs reading from the reader
+   */
+  private RuleResult searchResponse(final String hexDigest, final LineNumberReader reader) throws IOException
+  {
+    String line;
+    final Pattern p = Pattern.compile("^(" + hexDigest.substring(PREFIX_LENGTH) + "):(\\d+)\\s*$");
```

Comment:
I don't believe substring matching is correct here. According to https://haveibeenpwned.com/API/v2#PwnedPasswords:

> When a password hash with the same first 5 characters is found in the Pwned Passwords repository, the API will respond with an HTTP 200 and include the suffix of every hash beginning with the specified prefix, followed by a count of how many times it appears in the data set. The API consumer can then search the results of the response for the presence of their source hash and if not found, the password does not exist in the data set.

I take that to mean the _entire_ source has should be matched via simple startsWith semantics.

Reply:
I misunderstood both your code and the data returned. I read your code as only examining the _start_ of data, but I see it actually matches the _suffix_ of data, which is what is returned by the API. Thus my concerns were off base.